model/src/application/interaction/application_command/data/mod.rs

```diff
@@ -125,24 +81,250 @@ impl Serialize for CommandDataOption {
+                        Ok(Some(key)) => key,
+                        Ok(None) => break,
+                        Err(why) => {
+                            map.next_value::<IgnoredAny>()?;
+
+                            tracing::trace!("ran into an unknown key: {:?}", why);
+
+                            continue;
+                        }
+                    };
+
+                    match key {
+                        Fields::Name => {
+                            if name_opt.is_some() {
+                                return Err(DeError::duplicate_field("name"));
+                            }
+
+                            name_opt = Some(map.next_value()?);
+                        }
+                        Fields::Type => {
+                            if kind_opt.is_some() {
+                                return Err(DeError::duplicate_field("type"));
+                            }
+
+                            kind_opt = Some(map.next_value()?);
+                        }
+                        Fields::Options => {
+                            if !options.is_empty() {
+                                return Err(DeError::duplicate_field("options"));
+                            }
+
+                            options = map.next_value()?;
+                        }
+                        Fields::Value => {
+                            if value_opt.is_some() {
+                                return Err(DeError::duplicate_field("value"));
+                            }
+
+                            value_opt = Some(map.next_value()?);
+                        }
+                    }
+                }
+
+                let name = name_opt.ok_or_else(|| DeError::missing_field("name"))?;
+                let kind = kind_opt.ok_or_else(|| DeError::missing_field("type"))?;
+
+                let value = match kind {
+                    CommandOptionType::Boolean => {
+                        let val = value_opt.ok_or_else(|| DeError::missing_field("value"))?;
+
+                        if let ValueEnvelope::Boolean(b) = val {
+                            CommandOptionValue::Boolean(b)
+                        } else {
+                            return Err(DeError::custom(&format!(
+                                "expected boolean got {:?}",
+                                val
+                            )));
+                        }
+                    }
+                    CommandOptionType::Channel => {
+                        let val = value_opt.ok_or_else(|| DeError::missing_field("value"))?;
+
+                        match val {
+                            ValueEnvelope::Integer(i) => {
+                                let id = ChannelId::new(i.try_into().map_err(|_| {
+                                    DeError::custom("could not convert signed integer to unsigned")
+                                })?)
+                                .ok_or_else(|| DeError::custom("channel id was zero"))?;
```

Comment:
i think this could be written more clearly as such, and note that we would need to instead import `TryFrom` at the top:
```suggestion
                                let id = u64::try_from(i)
                                    .map(ChannelId::new)
                                    .map_err(|_| {
                                        DeError::custom(
                                            "could not convert signed integer to unsigned",
                                        )
                                    })?
                                    .ok_or_else(|| DeError::custom("channel id was zero"))?;
```
this can be done for the three other times this pattern appears

Reply:
Looking at the code again I am actually not sure if that code is even needed so I will do a bit of testing to figure it out.